<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Biological Data Science Workshops</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>
<!-- Font Awesome -->
<!-- <link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" /> -->
<!-- <script src="https://use.fontawesome.com/54ee8c2dfd.js"></script> -->

<!-- Google fonts -->
<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,400italic,700,700italic|Oswald:400,700' rel='stylesheet' type='text/css'>

<!-- Favicon -->
<link rel="shortcut icon" type="image/x-icon" href="img/favicon.ico">

<!-- Google analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-8298649-8', 'auto');
  ga('send', 'pageview');
</script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" type="text/css" />
<link rel="stylesheet" href="styles.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">BIODATASCI</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="setup.html">
    <span class="fa fa-cog"></span>
     
    Setup
  </a>
</li>
<li>
  <a href="data.html">
    <span class="fa fa-download"></span>
     
    Data
  </a>
</li>
<li>
  <a href="thriv.html">
    <span class="fa fa-graduation-cap"></span>
     
    Syllabus
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-university"></span>
     
    Course Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="thriv.html">
        <span class="fa fa-graduation-cap"></span>
         
        Syllabus
      </a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">--- Lessons ---</li>
    <li>
      <a href="r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="r-dataframes.html">Data Frames</a>
    </li>
    <li>
      <a href="r-dplyr-yeast.html">Data Manipulation</a>
    </li>
    <li>
      <a href="r-tidy.html">Tidying data</a>
    </li>
    <li>
      <a href="r-viz-gapminder.html">Data Visualization</a>
    </li>
    <li>
      <a href="r-refresher-tidy-eda.html">Refresher: Tidy EDA</a>
    </li>
    <li>
      <a href="r-rmarkdown.html">Reproducible Research &amp; RMarkdown</a>
    </li>
    <li>
      <a href="r-stats.html">Essential Statistics</a>
    </li>
    <li>
      <a href="r-survival.html">Survival Analysis</a>
    </li>
    <li>
      <a href="r-predictive-modeling.html">Predictive Analytics &amp; Forecasting Influenza</a>
    </li>
    <li>
      <a href="r-textmining.html">Text Mining</a>
    </li>
    <li class="dropdown-header">--- Homework ---</li>
    <li>
      <a href="r-dplyr-homework.html">Data Manipulation</a>
    </li>
    <li>
      <a href="r-viz-homework.html">Data Visualization</a>
    </li>
    <li>
      <a href="r-rmarkdown-homework.html">Reproducible Research &amp; RMarkdown</a>
    </li>
    <li>
      <a href="r-stats-homework.html">Essential Statistics</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="people.html">People</a>
    </li>
    <li>
      <a href="help.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/thriv/biodatasci">Source code for this site</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Biological Data Science Workshops</h1>

</div>


<p>This is a seven-part series of courses introducing the essentials of biomedical data science using R. This class introduces methods, tools, and software for reproducibly managing, manipulating, analyzing, and visualizing large-scale biological data using the R statistical computing environment. The workshops also cover essential statistical analysis, and advanced topics including survival analysis, predictive modeling, forecasting, and text mining.</p>
<!-- **Please see the _[THRIV syllabus](thriv.html)_ for more information.**  -->
<!-- **[Click here to download the entire course archive as a zip file.](https://github.com/thriv/biodatasci2018/archive/gh-pages.zip)** Once downloaded and extracted, double-click the **`index.html`** file in the extracted folder to view the course material offline. -->
<!-- ### Course Material -->
<!-- #### Lesson transcripts & code -->
<!-- 1. [R Basics](r-basics.html) -->
<!-- 1. [Data Frames](r-dataframes.html) -->
<!-- 1. [Data Manipulation](r-dplyr-yeast.html) -->
<!-- 1. [Tidying data](r-tidy.html) -->
<!-- 1. [Data Visualization](r-viz-gapminder.html) -->
<!-- 1. [Refresher: Tidy EDA](r-viz-gapminder.html) -->
<!-- 1. [Reproducible Research & RMarkdown](r-rmarkdown.html) -->
<!-- 1. [Essential Statistics with R](r-stats.html) -->
<!-- 1. [Survival Analysis](r-survival.html) -->
<!-- 1. [Predictive Analytics & Forecasting](r-predictive-modeling.html) -->
<!-- 1. [Text Mining](r-textmining.html) -->
<!-- #### Homework assignments -->
<!-- 1. [Data Manipulation](r-dplyr-homework.html) -->
<!-- 1. [Data Visualization](r-viz-homework.html) -->
<!-- 1. [Reproducible Research & RMarkdown](r-rmarkdown-homework.html) -->
<!-- 1. [Essential Statistics](r-stats-homework.html) -->
<hr />
<div id="faq" class="section level3">
<h3>FAQ</h3>
<div id="whats-this-series-all-about" class="section level4">
<h4>What’s this series all about?</h4>
<p>This class introduces methods, tools, and software for reproducibly managing, manipulating, analyzing, and visualizing large-scale biomedical data. Specifically, the course introduces the R statistical computing environment and packages for manipulating and visualizing high-dimensional data, covers strategies for reproducible research, and culminates with predictive modeling and forecasting analyses of real public health data.</p>
<p><strong>This is not a <em>“Tool X”</em> or <em>“Software Y”</em> class.</strong> I want you to take away from this series the ability to use an extremely powerful scientific computing environment (R) to do many of the things that you’ll do <em>across study designs and disciplines</em> – managing, manipulating, visualizing, and analyzing large, sometimes high-dimensional data. Whether that data is gene expression data from yeast, microbial genomics data from <em>B. pertussis</em>, public health data from <a href="http://www.gapminder.org/">Gapminder</a>, RNA-seq data from humans, movie preference trends from Netflix, or truck routing data from FedEx, you’ll need the same computational know-how and data literacy to do the same kinds of basic tasks in each. I might show you how to use specific tools here and there (DESeq2 for RNA-seq analysis, ggtree for drawing phylogenetic trees, etc.), but these are not important – you probably won’t be using the same specific software or methods 10 years from now, but you’ll still use the same underlying data and computational foundation. <strong>That</strong> is the point of this series – to arm you with a basic foundation, and more importantly, to <strong>enable you to figure out how to use <em>this tool</em> or <em>that tool</em> on your own</strong>, when you need to.</p>
<p><strong><em>This is not a statistics class.</em></strong> There is a short lesson on <a href="r-stats.html">essential statistics using R</a> but this 3-hour lesson offers neither a comprehensive background on underlying theory nor in-depth coverage of implementation strategies using R. Some general knowledge of statistics and study design is helpful, but isn’t required for this course.</p>
</div>
<div id="what-are-the-pre-requisites" class="section level4">
<h4>What are the pre-requisites?</h4>
<p><em>There are none!</em></p>
<p>However, <strong><a href="setup.html">click here for instructions on setting up your computer</a>.</strong> Each class involves lots of hands-on practice coding, and you’ll need to download and install some free software <strong><em>prior to our first class</em></strong>. This may take up to an hour or so, and please do not hesitate to <a href="people.html">email me</a> <em>prior to the workshop</em> if you are having difficulty.</p>
<!-- ## Can I audit this course? -->
<!-- Yes! However, **_you will be expected to attend every class meeting, participate in coding exercises during class, and complete any and all assignments_**, just as if you are taking the course for credit.  -->
<!-- **_UPDATE Feb 9 2016_**: The class is currently full. -->
<!-- [Click here to register to request to audit](https://docs.google.com/forms/d/1tHO-X4DupnHgIEsUei0K3kX5_UfLRK-y2KfxmxC6Ux0/viewform). The first day of the course is Monday, Feb 15, 2016. One week prior to the course starting, I will allow anyone who's requested to audit into the course, giving priority to people registering for credit. There are still plenty of seats open, so good chances you'll be able to get in. -->
<!-- ## Where will the class meet? -->
<!-- Carter Classroom, Health Sciences Library room 1211. This is downstairs one floor to the right (across the hall from Stephen Turner's office). -->
</div>
<div id="do-i-need-a-laptop" class="section level4">
<h4>Do I need a laptop?</h4>
<p><strong><em>YES.</em></strong> You must have access to a computer on which you can install software. The class will be a mix of lecture, discussion, but primarily live coding. You must bring your laptop to each session. Bring your charging cable also. Please follow the <a href="setup.html">setup instructions</a> prior to the workshop.</p>
</div>
<div id="where-do-i-get-additional-help" class="section level4">
<h4>Where do I get additional help?</h4>
<p>Glad you asked! <a href="help.html">See here</a>.</p>
<hr />
<p><small><strong><em>Attribution</em>:</strong> Course material is inspired by and/or modified in part from <a href="http://stat545-ubc.github.io/">Jenny Bryan’s Stat 545 course</a>, <a href="http://software-carpentry.org/">Software Carpentry</a>, <a href="http://datacarpentry.org/">Data Carpentry</a>, <a href="http://varianceexplained.org/">David Robinson’s blog</a>, <a href="https://github.com/marschmi/NGS2015_RMarkdown_Reproducibility">Marian Schmidt’s MSU NGS Workshop</a>, <a href="http://biostat.mc.vanderbilt.edu/wiki/Main/DataSets?CGISESSID=10713f6d891653ddcbb7ddbdd9cffb79">Vanderbilt Department of BioStatistics Datasets</a>, the <a href="http://bioconductor.org/packages/release/bioc/html/ggtree.html">ggtree vignettes</a>, <a href="https://shiring.github.io/machine_learning/2016/11/27/flu_outcome_ML_post">Shirin Glander’s blog</a>, <a href="https://www.tidytextmining.com/"><em>Text Mining with R</em></a> by Julia Silge and David Robinson, and likely many others.</small></p>
</div>
</div>

<div class="footer">
This work is licensed under the  <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0 Creative Commons License</a>.<br>
<!-- For more information, visit <a href="http://data.hsl.virginia.edu/" target="_blank">data.hsl.virginia.edu</a>.<br> -->
<a href="https://twitter.com/strnr" target="_blank"><i class="fa fa-twitter fa-lg"></i></a>&nbsp;
<a href="https://github.com/thriv/biodatasci" target="_blank"><i class="fa fa-github fa-lg"></i></a>&nbsp;
</div>



</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
