<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Data Frames</title>

<script src="site_libs/header-attrs-2.9/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<!-- Font Awesome -->
<!-- <link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" /> -->
<!-- <script src="https://use.fontawesome.com/54ee8c2dfd.js"></script> -->

<!-- Google fonts -->
<link href='https://fonts.googleapis.com/css?family=Open+Sans:400,400italic,700,700italic|Oswald:400,700' rel='stylesheet' type='text/css'>

<!-- Favicon -->
<link rel="shortcut icon" type="image/x-icon" href="img/favicon.ico">

<!-- Google analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
  ga('create', 'UA-8298649-8', 'auto');
  ga('send', 'pageview');
</script>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" type="text/css" />
<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">BIODATASCI</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="setup.html">
    <span class="fa fa-cog"></span>
     
    Setup
  </a>
</li>
<li>
  <a href="data.html">
    <span class="fa fa-download"></span>
     
    Data
  </a>
</li>
<li>
  <a href="syllabus.html">
    <span class="fa fa-graduation-cap"></span>
     
    Syllabus
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-university"></span>
     
    Course Material
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="syllabus.html">
        <span class="fa fa-graduation-cap"></span>
         
        Syllabus
      </a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">--- Lessons ---</li>
    <li>
      <a href="r-basics.html">R Basics</a>
    </li>
    <li>
      <a href="r-dataframes.html">Data Frames</a>
    </li>
    <li>
      <a href="r-dplyr-yeast.html">Data Manipulation</a>
    </li>
    <li>
      <a href="r-tidy.html">Tidying data</a>
    </li>
    <li>
      <a href="r-viz-gapminder.html">Data Visualization</a>
    </li>
    <li>
      <a href="r-refresher-tidy-eda.html">Refresher: Tidy EDA</a>
    </li>
    <li>
      <a href="r-rmarkdown.html">Reproducible Research &amp; RMarkdown</a>
    </li>
    <li>
      <a href="r-stats.html">Essential Statistics</a>
    </li>
    <li>
      <a href="r-survival.html">Survival Analysis</a>
    </li>
    <li>
      <a href="r-predictive-modeling.html">Predictive Analytics &amp; Forecasting Influenza</a>
    </li>
    <li>
      <a href="r-textmining.html">Text Mining</a>
    </li>
    <li>
      <a href="r-rnaseq-airway.html">RNA-seq</a>
    </li>
    <li>
      <a href="r-ggtree.html">Phylogenetic trees with R</a>
    </li>
    <li class="dropdown-header">--- Homework ---</li>
    <li>
      <a href="r-dplyr-homework.html">Data Manipulation</a>
    </li>
    <li>
      <a href="r-viz-homework.html">Data Visualization</a>
    </li>
    <li>
      <a href="r-rmarkdown-homework.html">Reproducible Research &amp; RMarkdown</a>
    </li>
    <li>
      <a href="r-stats-homework.html">Essential Statistics</a>
    </li>
    <li>
      <a href="r-rnaseq-homework.html">RNA-seq</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-question fa-lg"></span>
     
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="people.html">People</a>
    </li>
    <li>
      <a href="help.html">Further resources</a>
    </li>
    <li>
      <a href="https://github.com/stephenturner/workshops">Source code for this site</a>
    </li>
  </ul>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Data Frames</h1>

</div>


<p>There are <em>lots</em> of different basic data structures in R. If you take any kind of longer introduction to R you’ll probably learn about arrays, lists, matrices, etc. Let’s skip straight to the data structure you’ll probably use most – the <strong>data frame</strong>. We use data frames to store heterogeneous tabular data in R: tabular, meaning that individuals or observations are typically represented in rows, while variables or features are represented as columns; heterogeneous, meaning that columns/features/variables can be different classes (on variable, e.g. age, can be numeric, while another, e.g., cause of death, can be text).</p>
<p><strong>This lesson assumes a <a href="r-basics.html">basic familiarity with R</a>.</strong></p>
<p><strong>Recommended reading:</strong> Review the <a href="http://r4ds.had.co.nz/tibbles.html#introduction-4"><em>Introduction</em> (10.1)</a> and <a href="http://r4ds.had.co.nz/tibbles.html#tibbles-vs.data.frame"><em>Tibbles vs. data.frame</em> (10.3)</a> sections of the <a href="http://r4ds.had.co.nz/tibbles.html"><strong><em>R for Data Science</em> book</strong></a>. We will initially be using the <code>read_*</code> functions from the <a href="http://readr.tidyverse.org/"><strong>readr</strong> package</a>. These functions load data into a <em>tibble</em> instead of R’s traditional data.frame. Tibbles are data frames, but they tweak some older behaviors to make life a little easier. These sections explain the few key small differences between traditional data.frames and tibbles.</p>
<div id="our-data" class="section level2">
<h2>Our data</h2>
<!-- There are some built-in data frames that ship with R that you'll often see people referencing in help forums or other places to demonstrate some functionality. The `mtcars` data frame has data extracted from the 1974 _Motor Trend_ magazine giving statistics about design and performance for 32 different vehicles. You can just type `mtcars` to look at the whole dataset. -->
<p>The data we’re going to look at is cleaned up version of a gene expression dataset from <a href="http://www.ncbi.nlm.nih.gov/pubmed/17959824">Brauer et al. Coordination of Growth Rate, Cell Cycle, Stress Response, and Metabolic Activity in Yeast (2008) <em>Mol Biol Cell</em> 19:352-367</a>. This data is from a gene expression microarray, and in this paper the authors are examining the relationship between growth rate and gene expression in yeast cultures limited by one of six different nutrients (glucose, leucine, ammonium, sulfate, phosphate, uracil). If you give yeast a rich media loaded with nutrients except restrict the supply of a <em>single</em> nutrient, you can control the growth rate to any rate you choose. By starving yeast of specific nutrients you can find genes that:</p>
<ol style="list-style-type: decimal">
<li><strong>Raise or lower their expression in response to growth rate</strong>. Growth-rate dependent expression patterns can tell us a lot about cell cycle control, and how the cell responds to stress. The authors found that expression of &gt;25% of all yeast genes is linearly correlated with growth rate, independent of the limiting nutrient. They also found that the subset of negatively growth-correlated genes is enriched for peroxisomal functions, and positively correlated genes mainly encode ribosomal functions.</li>
<li><strong>Respond differently when different nutrients are being limited</strong>. If you see particular genes that respond very differently when a nutrient is sharply restricted, these genes might be involved in the transport or metabolism of that specific nutrient.</li>
</ol>
<p>You can download the cleaned up version of the data at <a href="data.html">the link above</a>. The file is called <a href="data/brauer2007_tidy.csv"><strong>brauer2007_tidy.csv</strong></a>. Later on we’ll actually start with the original raw data (minimally processed) and manipulate it so that we can make it more amenable for analysis.</p>
</div>
<div id="reading-in-data" class="section level2">
<h2>Reading in data</h2>
<div id="dplyr-and-readr" class="section level3">
<h3>dplyr and readr</h3>
<p>There are some built-in functions for reading in data in text files. These functions are <em>read-dot-something</em> – for example, <code>read.csv()</code> reads in comma-delimited text data; <code>read.delim()</code> reads in tab-delimited text, etc. We’re going to read in data a little bit differently here using the <a href="http://readr.tidyverse.org/">readr</a> package. When you load the readr package, you’ll have access to very similar looking functions, named <em>read-underscore-something</em> – e.g., <code>read_csv()</code>. You have to have the readr package installed to access these functions. Compared to the base functions, they’re <em>much</em> faster, they’re good at guessing the types of data in the columns, they don’t do some of the other silly things that the base functions do. We’re going to use another package later on called <a href="https://cran.r-project.org/web/packages/dplyr/index.html">dplyr</a>, and if you have the dplyr package loaded as well, and you read in the data with readr, the data will display nicely.</p>
<p>First let’s load those packages.</p>
<pre class="r"><code>library(readr)
library(dplyr)</code></pre>
<p>If you see a warning that looks like this: <code>Error in library(packageName) : there is no package called 'packageName'</code>, then you don’t have the package installed correctly. See the <a href="setup.html">setup page</a>.</p>
</div>
<div id="read_csv" class="section level3">
<h3><code>read_csv()</code></h3>
<p>Now, let’s actually load the data. You can get help for the import function with <code>?read_csv</code>. When we load data we assign it to a variable just like any other, and we can choose a name for that data. Since we’re going to be referring to this data a lot, let’s give it a short easy name to type. I’m going to call it <code>ydat</code>. Once we’ve loaded it we can type the name of the object itself (<code>ydat</code>) to see it printed to the screen.</p>
<pre class="r"><code>ydat &lt;- read_csv(file=&quot;data/brauer2007_tidy.csv&quot;)
ydat</code></pre>
<pre><code>## # A tibble: 198,430 × 7
##    symbol systematic_name nutrient  rate expression bp           mf             
##    &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;    &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt;          
##  1 SFB2   YNL049C         Glucose   0.05      -0.24 ER to Golgi… molecular func…
##  2 &lt;NA&gt;   YNL095C         Glucose   0.05       0.28 biological … molecular func…
##  3 QRI7   YDL104C         Glucose   0.05      -0.02 proteolysis… metalloendopep…
##  4 CFT2   YLR115W         Glucose   0.05      -0.33 mRNA polyad… RNA binding    
##  5 SSO2   YMR183C         Glucose   0.05       0.05 vesicle fus… t-SNARE activi…
##  6 PSP2   YML017W         Glucose   0.05      -0.69 biological … molecular func…
##  7 RIB2   YOL066C         Glucose   0.05      -0.55 riboflavin … pseudouridylat…
##  8 VMA13  YPR036W         Glucose   0.05      -0.75 vacuolar ac… hydrogen-trans…
##  9 EDC3   YEL015W         Glucose   0.05      -0.24 deadenylyla… molecular func…
## 10 VPS5   YOR069W         Glucose   0.05      -0.16 protein ret… protein transp…
## # … with 198,420 more rows</code></pre>
<p>Take a look at that output. The nice thing about loading dplyr and reading in data with readr is that data frames are displayed in a much more friendly way. This dataset has nearly 200,000 rows and 7 columns. When you import data this way and try to display the object in the console, instead of trying to display all 200,000 rows, you’ll only see about 10 by default. Also, if you have so many columns that the data would wrap off the edge of your screen, those columns will not be displayed, but you’ll see at the bottom of the output which, if any, columns were hidden from view. If you want to see the whole dataset, there are two ways to do this. First, you can click on the name of the data.frame in the <strong>Environment</strong> panel in RStudio. Or you could use the <code>View()</code> function (<em>with a capital V</em>).</p>
<pre class="r"><code>View(ydat)</code></pre>
</div>
</div>
<div id="inspecting-data.frame-objects" class="section level2">
<h2>Inspecting data.frame objects</h2>
<div id="built-in-functions" class="section level3">
<h3>Built-in functions</h3>
<p>There are several built-in functions that are useful for working with data frames.</p>
<ul>
<li>Content:
<ul>
<li><code>head()</code>: shows the first few rows</li>
<li><code>tail()</code>: shows the last few rows</li>
</ul></li>
<li>Size:
<ul>
<li><code>dim()</code>: returns a 2-element vector with the number of rows in the first element, and the number of columns as the second element (the dimensions of the object)</li>
<li><code>nrow()</code>: returns the number of rows</li>
<li><code>ncol()</code>: returns the number of columns</li>
</ul></li>
<li>Summary:
<ul>
<li><code>colnames()</code> (or just <code>names()</code>): returns the column names</li>
<li><code>str()</code>: structure of the object and information about the class, length and content of each column</li>
<li><code>summary()</code>: works differently depending on what kind of object you pass to it. Passing a data frame to the <code>summary()</code> function prints out useful summary statistics about numeric column (min, max, median, mean, etc.)</li>
</ul></li>
</ul>
<pre class="r"><code>head(ydat)</code></pre>
<pre><code>## # A tibble: 6 × 7
##   symbol systematic_name nutrient  rate expression bp                      mf   
##   &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;    &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;                   &lt;chr&gt;
## 1 SFB2   YNL049C         Glucose   0.05      -0.24 ER to Golgi transport   mole…
## 2 &lt;NA&gt;   YNL095C         Glucose   0.05       0.28 biological process unk… mole…
## 3 QRI7   YDL104C         Glucose   0.05      -0.02 proteolysis and peptid… meta…
## 4 CFT2   YLR115W         Glucose   0.05      -0.33 mRNA polyadenylylation* RNA …
## 5 SSO2   YMR183C         Glucose   0.05       0.05 vesicle fusion*         t-SN…
## 6 PSP2   YML017W         Glucose   0.05      -0.69 biological process unk… mole…</code></pre>
<pre class="r"><code>tail(ydat)</code></pre>
<pre><code>## # A tibble: 6 × 7
##   symbol systematic_name nutrient  rate expression bp                      mf   
##   &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;    &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;                   &lt;chr&gt;
## 1 DOA1   YKL213C         Uracil     0.3       0.14 ubiquitin-dependent pr… mole…
## 2 KRE1   YNL322C         Uracil     0.3       0.28 cell wall organization… stru…
## 3 MTL1   YGR023W         Uracil     0.3       0.27 cell wall organization… mole…
## 4 KRE9   YJL174W         Uracil     0.3       0.43 cell wall organization… mole…
## 5 UTH1   YKR042W         Uracil     0.3       0.19 mitochondrion organiza… mole…
## 6 &lt;NA&gt;   YOL111C         Uracil     0.3       0.04 biological process unk… mole…</code></pre>
<pre class="r"><code>dim(ydat)</code></pre>
<pre><code>## [1] 198430      7</code></pre>
<pre class="r"><code>names(ydat)</code></pre>
<pre><code>## [1] &quot;symbol&quot;          &quot;systematic_name&quot; &quot;nutrient&quot;        &quot;rate&quot;           
## [5] &quot;expression&quot;      &quot;bp&quot;              &quot;mf&quot;</code></pre>
<pre class="r"><code>str(ydat)</code></pre>
<pre><code>## spec_tbl_df [198,430 × 7] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
##  $ symbol         : chr [1:198430] &quot;SFB2&quot; NA &quot;QRI7&quot; &quot;CFT2&quot; ...
##  $ systematic_name: chr [1:198430] &quot;YNL049C&quot; &quot;YNL095C&quot; &quot;YDL104C&quot; &quot;YLR115W&quot; ...
##  $ nutrient       : chr [1:198430] &quot;Glucose&quot; &quot;Glucose&quot; &quot;Glucose&quot; &quot;Glucose&quot; ...
##  $ rate           : num [1:198430] 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 ...
##  $ expression     : num [1:198430] -0.24 0.28 -0.02 -0.33 0.05 -0.69 -0.55 -0.75 -0.24 -0.16 ...
##  $ bp             : chr [1:198430] &quot;ER to Golgi transport&quot; &quot;biological process unknown&quot; &quot;proteolysis and peptidolysis&quot; &quot;mRNA polyadenylylation*&quot; ...
##  $ mf             : chr [1:198430] &quot;molecular function unknown&quot; &quot;molecular function unknown&quot; &quot;metalloendopeptidase activity&quot; &quot;RNA binding&quot; ...
##  - attr(*, &quot;spec&quot;)=
##   .. cols(
##   ..   symbol = col_character(),
##   ..   systematic_name = col_character(),
##   ..   nutrient = col_character(),
##   ..   rate = col_double(),
##   ..   expression = col_double(),
##   ..   bp = col_character(),
##   ..   mf = col_character()
##   .. )
##  - attr(*, &quot;problems&quot;)=&lt;externalptr&gt;</code></pre>
<pre class="r"><code>summary(ydat)</code></pre>
<pre><code>##     symbol          systematic_name      nutrient              rate      
##  Length:198430      Length:198430      Length:198430      Min.   :0.050  
##  Class :character   Class :character   Class :character   1st Qu.:0.100  
##  Mode  :character   Mode  :character   Mode  :character   Median :0.200  
##                                                           Mean   :0.175  
##                                                           3rd Qu.:0.250  
##                                                           Max.   :0.300  
##    expression         bp                 mf           
##  Min.   :-6.50   Length:198430      Length:198430     
##  1st Qu.:-0.29   Class :character   Class :character  
##  Median : 0.00   Mode  :character   Mode  :character  
##  Mean   : 0.00                                        
##  3rd Qu.: 0.29                                        
##  Max.   : 6.64</code></pre>
</div>
<div id="other-packages" class="section level3">
<h3>Other packages</h3>
<p>The <code>glimpse()</code> function is available once you load the <strong>dplyr</strong> library, and it’s like <code>str()</code> but its display is a little bit better.</p>
<pre class="r"><code>glimpse(ydat)</code></pre>
<pre><code>## Rows: 198,430
## Columns: 7
## $ symbol          &lt;chr&gt; &quot;SFB2&quot;, NA, &quot;QRI7&quot;, &quot;CFT2&quot;, &quot;SSO2&quot;, &quot;PSP2&quot;, &quot;RIB2&quot;, &quot;V…
## $ systematic_name &lt;chr&gt; &quot;YNL049C&quot;, &quot;YNL095C&quot;, &quot;YDL104C&quot;, &quot;YLR115W&quot;, &quot;YMR183C&quot;,…
## $ nutrient        &lt;chr&gt; &quot;Glucose&quot;, &quot;Glucose&quot;, &quot;Glucose&quot;, &quot;Glucose&quot;, &quot;Glucose&quot;,…
## $ rate            &lt;dbl&gt; 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, …
## $ expression      &lt;dbl&gt; -0.24, 0.28, -0.02, -0.33, 0.05, -0.69, -0.55, -0.75, …
## $ bp              &lt;chr&gt; &quot;ER to Golgi transport&quot;, &quot;biological process unknown&quot;,…
## $ mf              &lt;chr&gt; &quot;molecular function unknown&quot;, &quot;molecular function unkn…</code></pre>
<p>The <strong>skimr</strong> package has a nice function, skim, that provides summary statistics the user can skim quickly to understand your data. You can install it with <code>install.packages("skimr")</code> if you don’t have it already.</p>
<pre class="r"><code>library(skimr)
skim(ydat)</code></pre>
<table>
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td align="left">Name</td>
<td align="left">ydat</td>
</tr>
<tr class="even">
<td align="left">Number of rows</td>
<td align="left">198430</td>
</tr>
<tr class="odd">
<td align="left">Number of columns</td>
<td align="left">7</td>
</tr>
<tr class="even">
<td align="left">_______________________</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Column type frequency:</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">character</td>
<td align="left">5</td>
</tr>
<tr class="odd">
<td align="left">numeric</td>
<td align="left">2</td>
</tr>
<tr class="even">
<td align="left">________________________</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Group variables</td>
<td align="left">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table>
<thead>
<tr class="header">
<th align="left">skim_variable</th>
<th align="right">n_missing</th>
<th align="right">complete_rate</th>
<th align="right">min</th>
<th align="right">max</th>
<th align="right">empty</th>
<th align="right">n_unique</th>
<th align="right">whitespace</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">symbol</td>
<td align="right">47250</td>
<td align="right">0.76</td>
<td align="right">2</td>
<td align="right">9</td>
<td align="right">0</td>
<td align="right">4210</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">systematic_name</td>
<td align="right">0</td>
<td align="right">1.00</td>
<td align="right">5</td>
<td align="right">9</td>
<td align="right">0</td>
<td align="right">5536</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">nutrient</td>
<td align="right">0</td>
<td align="right">1.00</td>
<td align="right">6</td>
<td align="right">9</td>
<td align="right">0</td>
<td align="right">6</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">bp</td>
<td align="right">7663</td>
<td align="right">0.96</td>
<td align="right">7</td>
<td align="right">82</td>
<td align="right">0</td>
<td align="right">880</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">mf</td>
<td align="right">7663</td>
<td align="right">0.96</td>
<td align="right">11</td>
<td align="right">125</td>
<td align="right">0</td>
<td align="right">1085</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table>
<thead>
<tr class="header">
<th align="left">skim_variable</th>
<th align="right">n_missing</th>
<th align="right">complete_rate</th>
<th align="right">mean</th>
<th align="right">sd</th>
<th align="right">p0</th>
<th align="right">p25</th>
<th align="right">p50</th>
<th align="right">p75</th>
<th align="right">p100</th>
<th align="left">hist</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">rate</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0.18</td>
<td align="right">0.09</td>
<td align="right">0.05</td>
<td align="right">0.10</td>
<td align="right">0.2</td>
<td align="right">0.25</td>
<td align="right">0.30</td>
<td align="left">▇▅▅▅▅</td>
</tr>
<tr class="even">
<td align="left">expression</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0.00</td>
<td align="right">0.67</td>
<td align="right">-6.50</td>
<td align="right">-0.29</td>
<td align="right">0.0</td>
<td align="right">0.29</td>
<td align="right">6.64</td>
<td align="left">▁▁▇▁▁</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="accessing-variables-subsetting-data-frames" class="section level2">
<h2>Accessing variables &amp; subsetting data frames</h2>
<p>We can access individual variables within a data frame using the <code>$</code> operator, e.g., <code>mydataframe$specificVariable</code>. Let’s print out all the gene names in the data. Then let’s calculate the average expression across all conditions, all genes (using the built-in <code>mean()</code> function).</p>
<pre class="r"><code># display all gene symbols
ydat$symbol</code></pre>
<pre><code>##   [1] &quot;SFB2&quot;     NA         &quot;QRI7&quot;     &quot;CFT2&quot;     &quot;SSO2&quot;     &quot;PSP2&quot;    
##   [7] &quot;RIB2&quot;     &quot;VMA13&quot;    &quot;EDC3&quot;     &quot;VPS5&quot;     NA         &quot;AMN1&quot;    
##  [13] &quot;SCW11&quot;    &quot;DSE2&quot;     &quot;COX15&quot;    &quot;SPE1&quot;     &quot;MTF1&quot;     &quot;KSS1&quot;    
##  [19] NA         NA         &quot;YAP7&quot;     NA         &quot;YVC1&quot;     &quot;CDC40&quot;   
##  [25] NA         &quot;RMD1&quot;     &quot;PCL6&quot;     &quot;AI4&quot;      &quot;GGC1&quot;     &quot;SUL1&quot;    
##  [31] &quot;RAD57&quot;    NA         &quot;PER1&quot;     &quot;YHC3&quot;     &quot;SGE1&quot;     &quot;HNM1&quot;    
##  [37] &quot;SWI1&quot;     &quot;NAM8&quot;     NA         &quot;BGL2&quot;     &quot;ACT1&quot;     NA        
##  [43] &quot;SFL1&quot;     &quot;OYE3&quot;     &quot;MMP1&quot;     &quot;MHT1&quot;     &quot;SUL2&quot;     &quot;IPP1&quot;    
##  [49] &quot;CWP1&quot;     &quot;SNF11&quot;    &quot;PEX25&quot;    &quot;ELO1&quot;     NA         &quot;CDC13&quot;   
##  [55] &quot;FKH1&quot;     &quot;SWD1&quot;     NA         &quot;HOF1&quot;     &quot;HOC1&quot;     &quot;BNI5&quot;    
##  [61] &quot;CSN12&quot;    &quot;PGS1&quot;     &quot;MLP2&quot;     &quot;HRP1&quot;     NA         &quot;SEC39&quot;   
##  [67] &quot;ECM31&quot;    NA         NA         &quot;ADE4&quot;     &quot;ABC1&quot;     &quot;DLD2&quot;    
##  [73] &quot;PHA2&quot;     NA         &quot;HAP3&quot;     &quot;MRPL23&quot;   NA         NA        
##  [79] &quot;MRPL16&quot;   NA         NA         NA         NA         &quot;AI3&quot;     
##  [85] &quot;COX1&quot;     NA         &quot;VAR1&quot;     &quot;COX3&quot;     &quot;COX2&quot;     &quot;AI5_BETA&quot;
##  [91] &quot;AI2&quot;      NA         NA         &quot;GPI18&quot;    &quot;COS9&quot;     NA        
##  [97] NA         &quot;PRP46&quot;    &quot;XDJ1&quot;     &quot;SLG1&quot;     &quot;MAM3&quot;     &quot;AEP1&quot;    
## [103] &quot;UGO1&quot;     NA         &quot;RSC2&quot;     &quot;YAP1801&quot;  &quot;ZPR1&quot;     &quot;BCD1&quot;    
## [109] &quot;UBP10&quot;    &quot;SLD3&quot;     &quot;RLF2&quot;     &quot;LRO1&quot;     NA         &quot;ITR2&quot;    
## [115] &quot;ABP140&quot;   &quot;STT3&quot;     &quot;PTC2&quot;     &quot;STE20&quot;    &quot;HRD3&quot;     &quot;CWH43&quot;   
## [121] &quot;ASK10&quot;    &quot;MPE1&quot;     &quot;SWC3&quot;     &quot;TSA1&quot;     &quot;ADE17&quot;    &quot;GFD2&quot;    
## [127] &quot;PXR1&quot;     NA         &quot;BUD14&quot;    &quot;AUS1&quot;     &quot;NHX1&quot;     &quot;NTE1&quot;    
## [133] NA         &quot;KIN3&quot;     &quot;BUD4&quot;     &quot;SLI15&quot;    &quot;PMT4&quot;     &quot;AVT5&quot;    
## [139] &quot;CHS2&quot;     &quot;GPI13&quot;    &quot;KAP95&quot;    &quot;EFT2&quot;     &quot;EFT1&quot;     &quot;GAS1&quot;    
## [145] &quot;CYK3&quot;     &quot;COQ2&quot;     &quot;PSD1&quot;     NA         &quot;PAC1&quot;     &quot;SUR7&quot;    
## [151] &quot;RAX1&quot;     &quot;DFM1&quot;     &quot;RBD2&quot;     NA         &quot;YIP4&quot;     &quot;SRB2&quot;    
## [157] &quot;HOL1&quot;     &quot;MEP3&quot;     NA         &quot;FEN2&quot;     NA         &quot;RFT1&quot;    
## [163] NA         &quot;MCK1&quot;     &quot;GPI10&quot;    &quot;APT1&quot;     NA         NA        
## [169] &quot;CPT1&quot;     &quot;ERV29&quot;    &quot;SFK1&quot;     NA         &quot;SEC20&quot;    &quot;TIR4&quot;    
## [175] NA         NA         &quot;ARC35&quot;    &quot;SOL1&quot;     &quot;BIO2&quot;     &quot;ASC1&quot;    
## [181] &quot;RBG1&quot;     &quot;PTC4&quot;     NA         &quot;OXA1&quot;     &quot;SIT4&quot;     &quot;PUB1&quot;    
## [187] &quot;FPR4&quot;     &quot;FUN12&quot;    &quot;DPH2&quot;     &quot;DPS1&quot;     &quot;DLD1&quot;     &quot;ASN2&quot;    
## [193] &quot;TRM9&quot;     &quot;DED81&quot;    &quot;SRM1&quot;     &quot;SAM50&quot;    &quot;POP2&quot;     &quot;FAA4&quot;    
## [199] NA         &quot;CEM1&quot;    
##  [ reached getOption(&quot;max.print&quot;) -- omitted 198230 entries ]</code></pre>
<pre class="r"><code>#mean expression
mean(ydat$expression)</code></pre>
<pre><code>## [1] 0.00337</code></pre>
<p>Now that’s not too interesting. This is the average gene expression across all genes, across all conditions. The data is actually scaled/centered around zero:</p>
<p><img src="r-dataframes_files/figure-html/histogram_expression_values-1.png" width="672" /></p>
<p>We might be interested in the average expression of genes with a particular biological function, and how that changes over different growth rates restricted by particular nutrients. This is the kind of thing we’re going to do in the next section.</p>
<hr />
<p><strong>EXERCISE 1</strong></p>
<ol style="list-style-type: decimal">
<li>What’s the standard deviation expression (hint: get help on the <code>sd</code> function with <code>?sd</code>).</li>
<li>What’s the range of rate represented in the data? (hint: <code>range()</code>).</li>
</ol>
<hr />
</div>
<div id="bonus-preview-to-advanced-manipulation" class="section level2">
<h2>BONUS: Preview to advanced manipulation</h2>
<p>What if we wanted show the mean expression, standard deviation, and correlation between growth rate and expression, separately for each limiting nutrient, separately for each gene, for all genes involved in the leucine biosynthesis pathway?</p>
<pre class="r"><code>ydat %&gt;% 
  filter(bp==&quot;leucine biosynthesis&quot;) %&gt;% 
  group_by(nutrient, symbol) %&gt;% 
  summarize(mean=mean(expression), sd=sd(expression), r=cor(rate, expression))</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">nutrient</th>
<th align="left">symbol</th>
<th align="right">mean</th>
<th align="right">sd</th>
<th align="right">r</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Ammonia</td>
<td align="left">LEU1</td>
<td align="right">-0.82</td>
<td align="right">0.39</td>
<td align="right">0.66</td>
</tr>
<tr class="even">
<td align="left">Ammonia</td>
<td align="left">LEU2</td>
<td align="right">-0.54</td>
<td align="right">0.38</td>
<td align="right">-0.19</td>
</tr>
<tr class="odd">
<td align="left">Ammonia</td>
<td align="left">LEU4</td>
<td align="right">-0.37</td>
<td align="right">0.56</td>
<td align="right">-0.67</td>
</tr>
<tr class="even">
<td align="left">Ammonia</td>
<td align="left">LEU9</td>
<td align="right">-1.01</td>
<td align="right">0.64</td>
<td align="right">0.87</td>
</tr>
<tr class="odd">
<td align="left">Glucose</td>
<td align="left">LEU1</td>
<td align="right">-0.55</td>
<td align="right">0.41</td>
<td align="right">0.98</td>
</tr>
<tr class="even">
<td align="left">Glucose</td>
<td align="left">LEU2</td>
<td align="right">-0.39</td>
<td align="right">0.33</td>
<td align="right">0.90</td>
</tr>
<tr class="odd">
<td align="left">Glucose</td>
<td align="left">LEU4</td>
<td align="right">1.09</td>
<td align="right">1.01</td>
<td align="right">-0.97</td>
</tr>
<tr class="even">
<td align="left">Glucose</td>
<td align="left">LEU9</td>
<td align="right">-0.16</td>
<td align="right">0.35</td>
<td align="right">0.35</td>
</tr>
<tr class="odd">
<td align="left">Leucine</td>
<td align="left">LEU1</td>
<td align="right">2.70</td>
<td align="right">1.08</td>
<td align="right">-0.95</td>
</tr>
<tr class="even">
<td align="left">Leucine</td>
<td align="left">LEU2</td>
<td align="right">0.28</td>
<td align="right">1.16</td>
<td align="right">-0.97</td>
</tr>
<tr class="odd">
<td align="left">Leucine</td>
<td align="left">LEU4</td>
<td align="right">0.80</td>
<td align="right">1.06</td>
<td align="right">-0.97</td>
</tr>
<tr class="even">
<td align="left">Leucine</td>
<td align="left">LEU9</td>
<td align="right">0.39</td>
<td align="right">0.18</td>
<td align="right">-0.77</td>
</tr>
<tr class="odd">
<td align="left">Phosphate</td>
<td align="left">LEU1</td>
<td align="right">-0.43</td>
<td align="right">0.27</td>
<td align="right">0.95</td>
</tr>
<tr class="even">
<td align="left">Phosphate</td>
<td align="left">LEU2</td>
<td align="right">-0.26</td>
<td align="right">0.19</td>
<td align="right">0.70</td>
</tr>
<tr class="odd">
<td align="left">Phosphate</td>
<td align="left">LEU4</td>
<td align="right">-0.99</td>
<td align="right">0.11</td>
<td align="right">0.24</td>
</tr>
<tr class="even">
<td align="left">Phosphate</td>
<td align="left">LEU9</td>
<td align="right">-1.12</td>
<td align="right">0.53</td>
<td align="right">0.90</td>
</tr>
<tr class="odd">
<td align="left">Sulfate</td>
<td align="left">LEU1</td>
<td align="right">-1.17</td>
<td align="right">0.34</td>
<td align="right">0.98</td>
</tr>
<tr class="even">
<td align="left">Sulfate</td>
<td align="left">LEU2</td>
<td align="right">-0.96</td>
<td align="right">0.30</td>
<td align="right">0.57</td>
</tr>
<tr class="odd">
<td align="left">Sulfate</td>
<td align="left">LEU4</td>
<td align="right">-0.24</td>
<td align="right">0.43</td>
<td align="right">-0.60</td>
</tr>
<tr class="even">
<td align="left">Sulfate</td>
<td align="left">LEU9</td>
<td align="right">-1.24</td>
<td align="right">0.55</td>
<td align="right">0.99</td>
</tr>
<tr class="odd">
<td align="left">Uracil</td>
<td align="left">LEU1</td>
<td align="right">-0.74</td>
<td align="right">0.73</td>
<td align="right">0.89</td>
</tr>
<tr class="even">
<td align="left">Uracil</td>
<td align="left">LEU2</td>
<td align="right">0.18</td>
<td align="right">0.13</td>
<td align="right">-0.07</td>
</tr>
<tr class="odd">
<td align="left">Uracil</td>
<td align="left">LEU4</td>
<td align="right">-0.65</td>
<td align="right">0.44</td>
<td align="right">0.77</td>
</tr>
<tr class="even">
<td align="left">Uracil</td>
<td align="left">LEU9</td>
<td align="right">-1.02</td>
<td align="right">0.91</td>
<td align="right">0.94</td>
</tr>
</tbody>
</table>
<p>Neat eh? We’ll learn how to do that in the advanced manipulation with dplyr lesson.</p>
<!-- Neat eh? Let's learn how to do that in the [advanced manipulation with dplyr lesson](r-dplyr-yeast.html). -->
</div>

<div class="footer">
This work is licensed under the  <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0 Creative Commons License</a>.<br>
<!-- For more information, visit <a href="http://data.hsl.virginia.edu/" target="_blank">data.hsl.virginia.edu</a>.<br> -->
<a href="https://twitter.com/strnr" target="_blank"><i class="fa fa-twitter fa-lg"></i></a>&nbsp;
<a href="https://github.com/thriv/biodatasci" target="_blank"><i class="fa fa-github fa-lg"></i></a>&nbsp;
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
